### ISSUE TITLE
Warnings from Fluentd metrics pods `[warn]: #0 HTTP status code 404`

### ISSUE THREAD

**Issue by astencel-sumo** [SumoLogic/sumologic-kubernetes-collection#1613]
**Describe the bug**
When running the collection, the Fluentd metrics pods output a lot of warnings like the below:

```console
$ kubectl -n nite-collections logs -f nite-collection-sumologic-fluentd-metrics-7
2021-05-25 13:22:22 +0000 [warn]: #0 HTTP status code 404, pods "nite-prometheus-kube-state-metrics-56cbbb5f5c-7w8tf" not found for GET /api/v1/namespaces/nite-collections/pods/nite-prometheus-kube-state-metrics-56cbbb5f5c-7w8tf
2021-05-25 13:22:22 +0000 [warn]: #0 HTTP status code 404, pods "nite-prometheus-kube-state-metrics-56cbbb5f5c-7w8tf" not found for GET /api/v1/namespaces/nite-monitoring/pods/nite-prometheus-kube-state-metrics-56cbbb5f5c-7w8tf
2021-05-25 13:22:22 +0000 [warn]: #0 HTTP status code 404, pods "nite-prometheus-kube-state-metrics-56cbbb5f5c-7w8tf" not found for GET /api/v1/namespaces/nite-live-stream/pods/nite-prometheus-kube-state-metrics-56cbbb5f5c-7w8tf
```

There doesn't seem to be anything wrong with the collection, but it would be nice to know why the logs are there and can we get rid of them. 

If you look closely, it's the same pod being requested for different namespaces, which is admittedly weird.

**To Reproduce**
Run collection in a cluster. Look at Fluentd metrics logs.

**Expected behavior**
No `HTTP status code 404` warnings.

**Environment (please complete the following information):**
 - **Collection version**: 2.1.1
 - **Kubernetes version**: 1.19
 - **Cloud provider**: EKS
 - **Others**:

**Comment by astencel-sumo** [comment]
There are actually two different issues in play here, both coming from the `enhance_k8s_metadata` plugin in https://github.com/SumoLogic/sumologic-kubernetes-fluentd.

# First issue - misusing `pod` and `namespace` labels for some metrics

This first issue manifests itself in multiple requests to API server having a common pod name but many different namespaces, like in the example above. This can be easily noticed for `kube-state-metrics` pods.

The cause of this issue is that for some metrics generated by `kube-state-metrics`, it doesn't make sense to use `pod` and `namespace` as a pair identifying the pod that the metric is about. For example, metrics that describe deployments have `deployment` and `namespace` set to the name of the deployment described by the metric and the namespace the deployment lives in. Then Prometheus adds the `pod` label to the metric that describes the name of the `kube-state-metrics` pod that generated this metric. In effect a metric is sent from Prometheus to Fluentd with the following metadata: 
- `deployment` - that contains the name of the deployment described by the metric,
- `namespace` - that contains the name of the namespace where that _deployment_ exists, 
- `pod` - name of the `kube-state-metrics` pod that generated this metric.

As can be seen from this data, taking `pod` and `namespace` as a pair does make sense. The `kube-state-metrics` pods generate metrics about every Kubernetes object (deployment, statefulset, daemonset, etc.) in every namespace. This causes the Fluentd plugin to try to retrieve data about `kube-state-metrics` pod in every namespace.

# Second issue - the cache is never cleaned from old pods

With the default configuration in v2.1.1 (up to and including v2.1.7 at the moment of writing), the plugin's pod metadata cache never goes stale. This means that if keeps metadata for pods that may be long gone and this stale data is never removed from cache (unless replaced with newer data when maximum cache size is reached, default limit 10000 entries).

This is especially a problem in clusters with high pod churn.

**Comment by astencel-sumo** [comment]
The `Second issue` is mostly fixed in [v2.3.0](https://github.com/SumoLogic/sumologic-kubernetes-collection/blob/release-v2.3/CHANGELOG.md#v230---2021-12-8) by removing pods from cache if the API server returns `404 Not Found` during cache refresh.

This was fixed in Fluentd image [v1.12.2-sumo-10](https://github.com/SumoLogic/sumologic-kubernetes-fluentd/releases/tag/v1.12.2-sumo-10).

**Comment by astencel-sumo** [comment]
The `First issue` does not in fact occur in a default installation of the collection chart. This has been fixed in September 2020 in https://github.com/SumoLogic/sumologic-kubernetes-collection/pull/878. When using a non-standard Prometheus remote write configuration for kube-state-metrics metrics, it is important to apply a similar relabeling change, changing the name of the `pod` label for non-pod metrics or just dropping the label altogether.
